File: Game_Of_Life_Python/Organisms/Animals/fox.py
from Organisms.organism import OrganismInitialData
from Organisms.animal import Animal
import random
from Organisms.animal import MoveDirection


class Fox(Animal):

    static_counter = 0

    def __init__(self, name, row, column, given_strength=-1, given_initiative=-1):
        strength = OrganismInitialData[name]["strength"]
        initiative = OrganismInitialData[name]["initiative"]
        character = OrganismInitialData[name]["character"]
        if given_strength != -1:
            strength = given_strength
        if given_initiative != -1:
            initiative = given_initiative
        super().__init__(strength, initiative, name, character, row, column, "fox.png")
        Fox.static_counter += 1

    def fox_can_go(self, character):
        return character == 't' or character == 'G' or character == 'U' or character == 'S' or character == 'e'

    def action(self, grid_board):

        self.previous_row = self.row
        self.previous_column = self.column

        rand = random.randint(1, 4)
        indexes = [i % 4 + 1 for i in range(rand, rand + 4)]
        moved = False
        for index in indexes:

            dir = MoveDirection(index)

            if dir == MoveDirection.LEFT and self.column > 0:
                if self.fox_can_go(grid_board[self.row][self.column - 1]):
                    self.organism_go_left()
                    moved = True
            elif dir == MoveDirection.RIGHT and self.column < len(grid_board[0]) - 1:
                if self.fox_can_go(grid_board[self.row][self.column + 1]):
                    self.organism_go_right()
                    moved = True
            elif dir == MoveDirection.TOP and self.row > 0:
                if self.fox_can_go(grid_board[self.row - 1][self.column]):
                    self.organism_go_top()
                    moved = True
            elif dir == MoveDirection.BOTTOM and self.row < len(grid_board) - 1:
                if self.fox_can_go(grid_board[self.row + 1][self.column]):
                    self.organism_go_bottom()
                    moved = True
            if moved:
                self.print_to_journal(
                    f"{self.name} from ({self.row}, {self.column}) to ")
                self.print_to_journal(f"""({self.row}, {self.column})\n""")
                return

        self.print_to_journal(f"""{self.name} stays at ({
                              self.row}, {self.column})\n""")

    def collision(self):
        pass

    def get_static_counter(self):
        return Fox.static_counter

File: GameOfLife_Cpp/source/src/fox.cpp
#include "fox.hpp"
#include <iostream>
#include <cassert>

int Fox::FOX_STATIC_COUNTER = 0;

Fox::Fox(int row, int column)
    : Animal(3, 7, 0, "fox", 'f', row, column, "fox.png", OrganismType::FOX)
{
    FOX_STATIC_COUNTER++;
}

Fox::Fox(int row, int column, int strength, int initiative, int age)
    : Animal(strength, initiative, age, "fox", 'f', row, column, "fox.png", OrganismType::FOX)
{
    FOX_STATIC_COUNTER++;
}

Fox::~Fox()
{
    FOX_STATIC_COUNTER--;
}

ActionResult Fox::action(std::vector<std::vector<char>>& grid_board)
{
    previous_row = row;
    previous_column = column;
    const int DIRECTIONS_AMOUNT = 4;
    int random_start = rand() % DIRECTIONS_AMOUNT + 1;
    int rand_array[DIRECTIONS_AMOUNT];
    
    int j = 0;
    for (int i = random_start; i < random_start + DIRECTIONS_AMOUNT; i++)
        rand_array[j++] = i % DIRECTIONS_AMOUNT + 1;

    Direction dir = Direction::NONE;
    for (int i = 0; i < DIRECTIONS_AMOUNT; i++)
    {
        dir = static_cast<Direction>(rand_array[i]);
        bool moved = false;
        switch (dir)
        {
            case Direction::LEFT:
                if (column > 0 && fox_can_go(grid_board[row][column - 1]))
                {
                    this->move_left();
                    moved = true;
                }
                break;
            case Direction::TOP:
                if (row > 0 && fox_can_go(grid_board[row - 1][column]))
                {
                    this->move_top();
                    moved = true;
                }
                break;
            case Direction::RIGHT:
                if (column < BOARD_WIDTH - 1 && fox_can_go(grid_board[row][column + 1]))
                {
                    this->move_right();
                    moved = true;
                }
                else
                {
                    if (column < BOARD_WIDTH - 1)
                        std::cout << "Cant move right now! : " << grid_board[row][column + 1] << std::endl;
                }
                break;
            case Direction::BOTTOM:
                if (row < BOARD_HEIGHT - 1 && fox_can_go(grid_board[row + 1][column]))
                {
                    this->move_bottom();
                    moved = true;
                }
                break;
            case Direction::NONE:
            default:
                break;
        }
        if (moved)
            break;
        dir = Direction::NONE;
    }

    // Fox had no option to move
    if (dir == Direction::NONE)
    {
        std::cout << this->get_name() << " at (" << this->row << ", " << this->column << ") has no cell to move\n";
        return ActionType::STAY;
    }

    std::cout << this->get_name() << " moves from (" << previous_row << ", " << previous_column << ") to (" << row << ", " << column << ")\n";
    this->default_grid_update(grid_board);

    return ActionResult(ActionType::MOVE);
}

CollisionResult Fox::collision(std::vector<std::vector<char>>& grid_board, std::vector<Organism*>& organisms, int current_index)
{
    return this->default_collision_animal(grid_board, organisms, current_index);
}

int Fox::get_static_counter()
{
    return FOX_STATIC_COUNTER;
}

bool Fox::fox_can_go(char character) const
{
    return character == 't' || character == 'G' || character == 'U' || character == 'S' || character == 'e' || character == 'f';
}